docs/source/notes/gradcheck.rst

```diff
@@ -0,0 +1,221 @@
+
+The constraints to satisfy complex differentiability (Cauchy-Riemann equations) are too restrictive for all real-valued loss functions, so we instead opted to use Wirtinger calculus.
+In a basic setting of Wirtinger calculus, the chain rule requires access to both the Wirtinger derivative (called :math:`W` below) and the Conjugate Wirtinger derivative (called :math:`CW` below).
+Both :math:`W` and :math:`CW` need to be propagated because in general, despite their name, one is not the complex conjugate of the other.
+
+To avoid having to propagate both values, for backward mode AD, we always work under the assumption that the function whose derivative is being calculated is either a real-valued function or is part of a bigger real-valued function. This assumption means that all the intermediary gradients we compute during the backward pass are also associated with real-valued functions.
+In practice, this assumption is not restrictive when doing optimization as such problem require real-valued objectives (as there is no natural ordering of the complex numbers).
+
+Under this assumption, we can show that :math:`W = CW^*` (we use :math:`*` to denote complex conjugation here) and so only one of the two values actually need to be "backwarded through the graph" as the other one can easily be recovered.
+To simplify internal computations, PyTorch uses :math:`2 * CW` as the value it backwards and returns when the user asks for gradients.
+Similarly to the real case, when the output is actually in :math:`\mathcal{R}^M`, backward mode AD does not compute :math:`2 * CW` but only :math:`v^T (2 * CW)` for a given vector :math:`v \in \mathcal{R}^M`.
+
+For forward mode AD, we use a similar logic, in this case, assuming that the function is part of a larger function whose input is in :math:`\mathcal{R}`. Under this assumption, we can make similar claim that every intermediary result corresponds to a function whose input is in :math:`\mathcal{R}` and in this case, we can show that :math:`W = CW` for the intermediary functions.
+To make sure the forward and backward mode compute the same quantities, the forward mode also computes :math:`2 * CW`.
+Similarly to the real case, when the input is actually in :math:`\mathcal{R}^N`, forward mode AD does not compute :math:`2 * CW` but only :math:`(2 * CW) u` for a given vector :math:`u \in \mathcal{R}^N`.
+
+
+Current backward mode gradcheck behavior
+----------------------------------------
+
+Functions in :math:`\mathcal{R}^N \to \mathcal{R}^M`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To test a function :math:`f: \mathcal{R}^N \to \mathcal{R}^M, x \to y`, we reconstruct the full Jacobian matrix :math:`J_f` of size :math:`M \times N` in two ways: analytical and numerical.
+The analytical version uses our backward mode AD while the numerical version uses finite difference.
+The two reconstructed Jacobian matrices are then compared elementwise for equality.
+
+Slow real input numerical evaluation
```

Comment:
We haven't defined what "slow gradcheck" is yet. For folks who aren't familiar with gradcheck, it's not obvious why this is "slow". We could define "slow gradcheck" or we write "full jacobian numeric evaluation" or something